winch/codegen/src/isa/x64/masm.rs

```diff
@@ -671,6 +671,27 @@ impl Masm for MacroAssembler {
         self.asm.setcc(kind, dst);
     }
 
+    fn float_cmp(&mut self, src1: Reg, src2: Reg, dst: Reg, kind: FloatCmpKind, size: OperandSize) {
+        // xmm_cmp_rr does not support ordered greater than comparisons so when
+        // performing a greater than comparison, swap the operands and use the
+        // corresponding less than operation instead.
+        let (src1, src2, kind) = match kind {
+            FloatCmpKind::Eq | FloatCmpKind::Ne | FloatCmpKind::Lt | FloatCmpKind::Le => {
+                (src1, src2, kind)
+            }
+            FloatCmpKind::Gt => (src2, src1, FloatCmpKind::Lt),
+            FloatCmpKind::Ge => (src2, src1, FloatCmpKind::Le),
+        };
+
+        self.asm.xmm_cmp_rr(src1, src2, kind, size);
+        // If comparison is true, src2 will be a 32 bit or 64 bit mask of all 1s.
+        self.asm.xmm_to_gpr(src2, dst, size, OperandSize::S32);
+        // If comparison is true, dst will now be `0x80000000`.
+        // Shifting 0x80000000 right by 31 bits yields 0x1.
+        self.asm
+            .shift_ir(31, dst, ShiftKind::ShrU, OperandSize::S32);
+    }
+
```

Comment:
When we first discussed the implementation of this operation, I didn't fully realize that this approach involves an extra move and a shift. By investigating a bit more, I've convinced myself that we can do better here by using `XmmCmpRmR` with the [`Ucomis{s,d}`](https://www.felixcloutier.com/x86/ucomiss) `SseOpcode`; so what I'm suggesting is:

- Rename this function to `float_cmp_with_set`, similar to what we have for int operations.
- Introduce a `ucomis` instruction at the `Assembler` layer. 
- Use `XmmCmpRmR` with `Ucomis,{s,d}`
- And use the `SETcc` instruction to set the comparison result in the `dst` register. 

I also realized that the `Inst::XmmRmRImm` is mostly useful with vector operands; with scalars most compilers use the ucomis{s,d} instructions.

Reply:
One challenge I've run into with `ucomis{s,d}` is that it seems the comparison instructions should be ordered (except for not equal).

https://github.com/sunfishcode/wasm-reference-manual/blob/master/WebAssembly.md#floating-point-equality says:
> This instruction corresponds to what is sometimes called "ordered and equal", or "oeq", in other languages.

And I've noticed the test suite fails when trying to use `ucomis{s,d}` to get the results.

Unfortunately I haven't figured out how to get `comis{s,d}` to work with what's defined in `cranelift/codegen/src/isa/x64/inst/emit.rs`.

Reply:
Hmm interesting, does the spec test suite fails in a way that hints that this issue is related to the use of `ucomis{s,d}`? As a comparison point, Cranelift generates the following x64 assembly for the function:

```wat
(module
  (func (param $x f32) (param $y f32) (result i32)
    (local.get 0)
    (local.get 1)
    f32.lt
  )
)
```

```asm
Disassembly of 20 bytes:
   0:   55                      pushq   %rbp
   1:   48 89 e5                movq    %rsp, %rbp
   4:   0f 2e c8                ucomiss %xmm0, %xmm1
   7:   41 0f 97 c1             seta    %r9b
   b:   41 0f b6 c1             movzbl  %r9b, %eax
   f:   48 89 ec                movq    %rbp, %rsp
  12:   5d                      popq    %rbp
  13:   c3                      retq
```

Reply:
Oh nvm, I see that you have specified that it fails on the `NaN` test cases.

Reply:
But it's still interesting to me, since according to how I understand the instruction definition it's supposed to handle `NaN`s correctly.

Reply:
I found having to check the parity flag for that surprising because my understanding is the parity flag is set based on whether the least significant bit is even or not. However, I found out that the FPU sets 4 condition flags (C0 through C3) and C2 is set when two numbers that are incomparable are compared. And that C2 is stored in the parity flag. So now that makes more sense to me and gives us a path forward.

Reply:
> But it's still interesting to me, since according to how I understand the instruction definition it's supposed to handle NaNs correctly.

`Ucomis{s,d}` is an unordered comparison and `f32.eq`/`f64.eq` are ordered comparisons.

It looks like the other comparisons are also ordered except for `f32.ne` and `f64.ne`.

I've got something that looks like:

```rust
        self.asm.ucomis(src1, src2, size);
        self.asm.setcc(kind, dst);
        if !matches!(kind, CmpKind::Ne) {
            let scratch = regs::scratch();
            self.asm.setcc(CmpKind::NotParity, scratch);
            self.asm.and_rr(scratch, dst, size);
        }
```
but I'm still having trouble getting the `ne` suite to pass on the NaN comparisons.

Reply:
> Ucomis{s,d} is an unordered comparison and f32.eq/f64.eq are ordered comparisons.

That's why it correctly handles NaNs (setting EFLAGS vs failing like `comis{s,d}`),  or is there anything else here that we aren't aligned on? 

Given that the Wasm spec states that all the other operations are ordered (which means that any comparison involves NaNs should be false) except for `{f32,f64}.ne` wouldn't that mean that for `ne`, instead of emitting an `and` with between the parity flag and equality flag, we'd need to emit an or?

```asm
   0:   55                      pushq   %rbp
   1:   48 89 e5                movq    %rsp, %rbp
   4:   0f 2e c1                ucomiss %xmm1, %xmm0
   7:   41 0f 9a c3             setp    %r11b
   b:   40 0f 95 c6             setne   %sil
   f:   41 09 f3                orl     %esi, %r11d
  12:   41 0f b6 c3             movzbl  %r11b, %eax
  16:   48 89 ec                movq    %rbp, %rsp
  19:   5d                      popq    %rbp
  1a:   c3                      retq
```
```wat
  (func (export "ne") (param $x f32) (param $y f32) (result i32) (f32.ne (local.get $x) (local.get $y)))
```

Reply:
> That's why it correctly handles NaNs (setting EFLAGS vs failing like comis{s,d}), or is there anything else here that we aren't aligned on?

Sorry my point was more that the comparison being ordered is ucomis on it's own isn't sufficient but we're agreed on that.

> Given that the Wasm spec states that all the other operations are ordered (which means that any comparison involves NaNs should be false) except for {f32,f64}.ne wouldn't that mean that for ne, instead of emitting an and with between the parity flag and equality flag, we'd need to emit an or?

Almost. For the `eq` case, it's an `and` between _not parity_ and the result and for `ne`, it's an `or` between _parity_ and the result. But now I have `(assert_return (invoke "lt" (f32.const -0x0p+0) (f32.const 0x1p-149)) (i32.const 1))` failing and I have to go.

Current code:

```rust
        self.asm.ucomis(src1, src2, size);
        self.asm.setcc(kind, dst);
        let scratch = regs::scratch();
        if !matches!(kind, CmpKind::Ne) {
            self.asm.setcc(CmpKind::NotParity, scratch);
            self.asm.and_rr(scratch, dst, size);
        } else {
            self.asm.setcc(CmpKind::Parity, scratch);
            self.asm.or_rr(scratch, dst, size);
        }
```

Reply:
Got it working!